third_party/maya/lib/usdMaya/xformStack.cpp

```diff
@@ -0,0 +1,748 @@
+                || opType == UsdGeomXformOp::TypeRotateYZX
+                || opType == UsdGeomXformOp::TypeRotateZXY
+                || opType == UsdGeomXformOp::TypeRotateZYX);
+    }
+
+    bool _isOneOrThreeAxisRotate(UsdGeomXformOp::Type opType)
+    {
+        return (opType == UsdGeomXformOp::TypeRotateXYZ
+                || opType == UsdGeomXformOp::TypeRotateXZY
+                || opType == UsdGeomXformOp::TypeRotateYXZ
+                || opType == UsdGeomXformOp::TypeRotateYZX
+                || opType == UsdGeomXformOp::TypeRotateZXY
+                || opType == UsdGeomXformOp::TypeRotateZYX
+                || opType == UsdGeomXformOp::TypeRotateX
+                || opType == UsdGeomXformOp::TypeRotateY
+                || opType == UsdGeomXformOp::TypeRotateZ);
+    }
+
+    IndexMap _buildInversionMap(
+            const std::vector<std::pair<size_t, size_t> >& inversionTwins)
+    {
+        IndexMap result;
+        for (const auto& twinPair : inversionTwins)
+        {
+            result[twinPair.first] = twinPair.second;
+            result[twinPair.second] = twinPair.first;
+        }
+        return result;
+    }
+
+    // Given a single index into ops, return the pair of
+    // indices, which is:
+    //     { opIndex, NO_INDEX }     if opIndex has no inversion twin
+    //     { opIndex, opIndexTwin }  if opIndex has an inversion twin, and opIndex < opIndexTwin
+    //     { opIndexTwin, opIndex }  if opIndex has an inversion twin, and opIndex > opIndexTwin
+    IndexPair
+    _makeInversionIndexPair(
+            size_t opIndex,
+            const IndexMap& inversionMap)
+    {
+        auto foundTwin = inversionMap.find(opIndex);
+        if (foundTwin == inversionMap.end())
+        {
+            return std::make_pair(opIndex, PxrUsdMayaXformStack::NO_INDEX);
+        }
+        else
+        {
+            const size_t twinOpIndex = foundTwin->second;
+            if (twinOpIndex >= opIndex)
+            {
+                return std::make_pair(opIndex, twinOpIndex);
+            }
+            else
+            {
+                return std::make_pair(twinOpIndex, opIndex);
+            }
+        }
+    }
+
+    TokenIndexPairMap _buildAttrNamesToIdxs(
+            const OpClassList& ops,
+            const IndexMap& inversionMap)
+    {
+        TokenIndexPairMap result;
+        for (size_t i = 0; i < ops.size(); ++i)
+        {
+            const PxrUsdMayaXformOpClassification& op = ops[i];
+            // Inverted twin pairs will always have same name, so can skip one
+            if (op.IsInvertedTwin()) continue;
+
+            for (auto attrName : op.CompatibleAttrNames())
+            {
+                auto indexPair = _makeInversionIndexPair(i, inversionMap);
+                // Insert, and check if it already existed
+                if (!result.insert({attrName, indexPair}).second)
+                {
+                    // The attrName was already in the lookup map...
```

Comment:
Our coding convention is to not use exceptions in our code, so we'd prefer not to throw here. If this is something that should never happen (i.e., it's an invariant in the code and not reliant on user input), we prefer to use TF_VERIFY to indicate that something is wrong, like:

```
TF_VERIFY(result.insert({attrName, indexPair}).second, 
          "AttrName %s already found in attrName lookup map", 
          attrName.GetText());
```
This will cause the error output to be emitted if the condition is false, but allows the code to continue executing so that it doesn't cause the program to crash.